Funcky/Extensions/ParseExtensions/ParseExtensions.DateTime.cs

```diff
@@ -0,0 +1,78 @@
+using System;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using Funcky.Monads;
+
+namespace Funcky.Extensions
+{
+    public static partial class ParseExtensions
+    {
+        [Pure]
+        public static Option<DateTime> ParseDateTimeOrNone(this string candidate)
+            => DateTime.TryParse(candidate, out var dateTime)
+                ? dateTime
+                : Option<DateTime>.None();
+
+#if READ_ONLY_SPAN_SUPPORTED
+        [Pure]
+        public static Option<DateTime> ParseDateTimeOrNone(this ReadOnlySpan<char> candidate)
+            => DateTime.TryParse(candidate, out var dateTime)
+                ? dateTime
+                : Option<DateTime>.None();
+#endif
+
+        [Pure]
+        public static Option<DateTime> ParseDateTimeOrNone(this string candidate, IFormatProvider provider, DateTimeStyles styles)
```

Comment:
The `provider` and `styles` parameters have default values on the original `TryParse` method. Do we want to have these too?

Reply:
Are you sure?
In the underlying implementation the last parameter is an out parameter, there cannot be a default parameter before that. I only checked the documentation though.

Reply:
Oh sorry I accidentally looked at the docs for `Parse` instead of `TryParse`. But there is another nullable param besides the input string: `provider`

Reply:
The parse overloads for string also do not have default providers, only the new ones for ReadOnlySpan<char> actually do that, it is really looks like an inconsistent mess.

Adding default parameters would also lead to a different number of overloads between DateTime.TryParse / and ParseDateTimeOrNone. I am not sure how to proceed.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.cs

Reply:
*sigh* What's more important to us? Internal consistency or consistency with .NET?

Given that the `ReadOnlySpan` overloads are newer (and there must have been a reason for the defaults), we could use the defaults from those for the `string` overloads too.
But then again those overloads are only needed because there's only one overload for `ReadOnlySpan` while for `string` there are three separate overloads.

Reply:
Per discussion IRL we decided to go with the "default parameters style" for all Parse methods.

Reply:
Fixed in 5365f56d8ebc6aa0d5671bc283d282148e241344